### DISCUSSION TITLE
How to pass data from filter hook to action hook?

### DISCUSSION THREAD

**Discussion by sarthakgaur** [directus/directus#13566] [Archive]
Title says it all.

**Comment by rijkvanzanten** [comment]
What do you mean pass from filter to action? Do you have an example of when you'd need do pass data from a filter to an action hook, or what the actual use case would be?

**Reply by sarthakgaur** [reply]
Let's say there are two tables: A and B. When a row from table A is deleted we have to update a field in a row in table B. In the `filter delete` hook of A, we do some computation based on the row that will get deleted. But, we can't perform an update in table B because the row in table A is not yet deleted. So, only in the `action delete` hook of A, we can perform that update. But, we can't perform any computation in the `action delete` hook of A as the row of A that we need for the computation is already deleted.

So, these are the steps we need to do:

1. Perform the computation in `filter delete` of A.
2. Pass that computed value to the `action delete` hook of A.
3. Perform the row update in table B in `action delete` hook of A using the passed computed value.

We have noticed that if we modify the payload in the `filter delete` hook of A, we get the modified payload in the `action delete` hook of A. So, we can use it as a way to pass data from filter hook to action hook. Should we go ahead with this approach?

**Reply by rijkvanzanten** [reply]
There's currently no official way to pass context from one to the other in this fashion. The payload approach could work, but would only work on deleting hooks, as it's normally used for the payload that's about to be saved to the database. You might be better off creating a dict outside of the hooks themselves to temporarily store this information, where you'd use the PK of the deleted item as the key in the dict, something like (pseudo code):

```js
module.exports = function registerHooks({ action, filter }) {
  const state = {};

  filter('items.delete', (input, { collection, keys }) => {
    state[collection + '_' + keys[0]] = 'my shared value';
  });

  action('items.delete', ({ collection, keys }) => {
    const sharedValue = state[collection + '_' + keys[0];

    // Do whatever with the shared value    

    delete state[collection + '_' + keys[0];
  });
}
```

**Reply by sarthakgaur** [reply]
> The payload approach could work, but would only work on deleting hooks, as it's normally used for the payload that's about to be saved to the database.

We tried it on `filter update` and `action update` hooks, too. It works fine in that case, too. I think we just have to add a payload field that doesn't conflict with the column names.

We also thought about the map approach that you suggested. But, we thought that this should be quite common that there must be a dedicated way of doing it.

**Reply by rijkvanzanten** [reply]
> I think we just have to add a payload field that doesn't conflict with the column names.

Fair! Though it's not excluded that there will be stricter validation in Directus itself in the future to prevent payloads with wrong column names to avoid unintended changes/wrong updates 🤔 

I think the map approach is the safest way right now.

> we thought that this should be quite common that there must be a dedicated way of doing it.

You're the first one to bring it up!
